File: src/examples/execute_traj.py
#!/usr/bin/env python

import rospy
from roscpp_initializer import roscpp_initializer
from aikidopy import SkeletonMarker, InteractiveMarkerViewer
from cozmopy import Cozmo, Waypoint
from math import pi 
from datetime import timedelta
import argparse

if __name__ == "__main__":
	parser = argparse.ArgumentParser()
	parser.add_argument('--mesh_dir', dest='mesh_dir', required = True,
	                     help='The path to the directory containing Cozmos meshes')
	args = parser.parse_args()

	# rospy.init_node does not initialize roscpp and if roscpp is not initialized
	# instanciating ros::NodeHandle will lead to a fatal error.
	# roscpp_initializer initializes roscpp and ros::NodeHandle in the background
	roscpp_initializer.roscpp_init("load_cozmo", [])
	rospy.init_node("load_cozmo")
	rate = rospy.Rate(10)

	topicName = "dart_markers"
	baseFrameName = "map"

	if not rospy.is_shutdown():
		cozmo = Cozmo(args.mesh_dir)
		skeleton = cozmo.getCozmoSkeleton();

		print("Starting viewer. Please subscribe to the {} InteractiveMarker" 
			  " topic in Rviz".format(topicName))

		viewer = InteractiveMarkerViewer(topicName, baseFrameName)
		
		cozmo_marker = viewer.addSkeletonMarker(skeleton)
		viewer.setAutoUpdate(True);

		waypoints = [
			Waypoint(0.0, 0.0, 0, 0),
		    Waypoint(0.3, 0.0, 0, 2),
		    Waypoint(0.3, 0.0, pi/2, 3),
		    Waypoint(0.3, 0.3, pi/2, 5),
		    Waypoint(0.3, 0.3, pi, 6),
		    Waypoint(0.0, 0.3, pi, 8),
		    Waypoint(0.0, 0.3, -pi/2, 9),
		    Waypoint(0.0, 0.0, -pi/2, 11),
		    Waypoint(0.0, 0.0, 0, 12),
		]

		traj = cozmo.createInterpolatedTraj(waypoints);
		cozmo.executeTrajectory(timedelta(milliseconds=6), traj)	

File: src/examples/execute_traj.cpp
#include "cozmo_description/cozmo.hpp"
#include "ros/ros.h"
#include <chrono>
#include <cstdlib>
#include <math.h>
#include "aikido/trajectory/Interpolated.hpp"
#include <aikido/rviz/InteractiveMarkerViewer.hpp>
#include "aikido/rviz/FrameMarker.hpp"

using Interpolated = aikido::trajectory::Interpolated;

static const std::string topicName("dart_markers");
static const std::string baseFrameName("base_link");

int main(int argc, char* argv[])
{
  if (argc < 2) {
    std::cerr << "Please include the path to the mesh directory" << std::endl;
    return 0;
  }

  const std::string mesh_dir = argv[1];
  libcozmo::Cozmo cozmo(mesh_dir);

  // Start the RViz viewer.
  std::cout << "Starting ROS node." << std::endl;
  ros::init(argc, argv, "load_cozmo");
  ros::NodeHandle nh("~");

  dart::dynamics::SkeletonPtr skeleton;
  skeleton = cozmo.getCozmoSkeleton();

  std::cout << "Starting viewer. Please subscribe to the '" << topicName
	    << "' InteractiveMarker topic in RViz." << std::endl;

  aikido::rviz::InteractiveMarkerViewer viewer(topicName, baseFrameName);
  viewer.addSkeletonMarker(skeleton);

  auto body = skeleton->getBodyNode(0);
  // Create x-axis
  auto frameShapeFrameX = dart::dynamics::SimpleFrame(body, "axis x");
  auto arrowX = std::make_shared<dart::dynamics::ArrowShape>(Eigen::Vector3d::Zero(), Eigen::Vector3d::UnitX());
  frameShapeFrameX.setShape(arrowX);

  auto frameShapeFrameY = dart::dynamics::SimpleFrame(body, "axis y");
  auto arrowY = std::make_shared<dart::dynamics::ArrowShape>(Eigen::Vector3d::Zero(), Eigen::Vector3d::UnitY());
  frameShapeFrameY.setShape(arrowY);

  auto frameShapeFrameZ = dart::dynamics::SimpleFrame(body, "axis z");
  auto arrowZ = std::make_shared<dart::dynamics::ArrowShape>(Eigen::Vector3d::Zero(), Eigen::Vector3d::UnitZ());
  frameShapeFrameZ.setShape(arrowZ);
  viewer.addFrame(&frameShapeFrameX);
  viewer.addFrame(&frameShapeFrameY);
  viewer.addFrame(&frameShapeFrameZ);
  // Eigen::Isometry3d tf = frame->getWorldTransform();
  // auto world = dart::dynamics::Frame::World();
  // dart::dynamics::SimpleFrame simple_frame(world, "target", tf);
  // viewer.addFrame(&simple_frame);

  viewer.setAutoUpdate(true);

  std::vector<libcozmo::Waypoint> waypoints = {
      {.x = 0.0, .y = 0.0, .th = 0, .t = 0},
      {.x = 0.3, .y = 0.0, .th = 0, .t = 2},
      {.x = 0.3, .y = 0.0, .th = M_PI/2, .t = 3},
      {.x = 0.3, .y = 0.3, .th = M_PI/2, .t = 5},
      {.x = 0.3, .y = 0.3, .th = M_PI, .t = 6},
      {.x = 0.0, .y = 0.3, .th = M_PI, .t = 8},
      {.x = 0.0, .y = 0.3, .th = -M_PI/2, .t = 9},
      {.x = 0.0, .y = 0.0, .th = -M_PI/2, .t = 11},
      {.x = 0.0, .y = 0.0, .th = 0, .t = 12}
  };

  std::shared_ptr<Interpolated> traj;
  traj = cozmo.createInterpolatedTraj(waypoints);

  std::chrono::milliseconds period(6);
  cozmo.executeTrajectory(period, traj);

  return 0;
}
